### PR TITLE
Feat/route refactoring

### PR DESCRIPTION
**Pull request by alecritson**
This PR is an initial Draft to route refactoring. A list of changes are as follows:

(All changes relate to the `routes` table unless specified)

- Drop the `path` column as this was confusing and didn't really serve a purpose
- `locale` has been dropped in favour of a `lanugage_id` relation to `languages`
- Added index to `element_type`
- We had separate routes for adding routes to different resources i.e. `product/{id}/routes` this has been removed in favour of a global `routes` endpoint for creation
- When creating a route you must specify `language_id` `element_id` and `element_type`. The `slug` and `element_type` columns must be unique when used together.
- A new `RouteResolver` class (naming?) has been added so that you can pass through a simplified `element_type` instead of the whole class name and it will be resolved into the fully qualified name. For example, `products` would resolve to `GetCandy\Api\Core\Products\Models\Product` which would be unrealistic to send up via Axios.
- The `RouteResolver` also exposes a function `addElementType($key, $value)` so any plugins can add their own and allow resolving.
- When searching for a route you are now required to supply `slug` and `element_type`, wasn't sure about `language_id` didn't think it was required?
- The signature of `RouteResource` has been updated to reflect these changes.

### CODE REVIEW

**Comment by glennjacobs** [comment]
In terms of uniqueness and language_id, we either make language_id part of the uniqueness or make language_id mandatory when searching for a route. It's either-or.

For the resolver, call it `AliasResolver` and make it generic so that other parts of GetCandy could potentially use it. And as you suggested, it could throw a `AliasResolutionException` if a match isn't found.

File: CHANGELOG.md [hunk 1]
```diff
@@ -37,7 +37,33 @@ You should update any calls to this endpoint if you rely on included resources.
...
+        include: 'element'
+    }
+})
+```
+This should now be changed to:
+```javascript
+const { data } = await axios('routes/search', {
+    params: {
+        slug: 'slug-for-the-product',
+        language_id: '6z8m9gmj',
```
Comment (glennjacobs) [inline CHANGELOG.md:61]: I don't think this is very developer-friendly. Would be better being the code.

File: database/migrations/2021_03_19_092900_add_unique_index_to_language_codes.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        });
+    }
+
+    /**
+     * Reverse the migrations.
+     */
+    public function down()
+    {
+        Schema::table('languages', function (Blueprint $table) {
+            $table->dropIndex(['code']);
```
Comment (glennjacobs) [inline database/migrations/2021_03_19_092900_add_unique_index_to_language_codes.php:25]: Pretty sure when dropping indexes you have to use the full index name.

Comment (alecritson) [reply database/migrations/2021_03_19_092900_add_unique_index_to_language_codes.php:25]: Not when you pass an array I believe

**Pull request merged**